pkg/tfshim/sdk-v2/upgrade_state.go

```diff
@@ -35,13 +39,18 @@ func upgradeResourceState(ctx context.Context, typeName string, p *schema.Provid
 		version, hasVersion = v, true
 	}
 
-	// Now, we perform the UpgradeResourceState operation by re-implementing TF's UpgradeResourceState.
+	json, err := ctyjson.Marshal(instanceState.RawState, res.CoreConfigSchema().ImpliedType())
```

Comment:
Eyes on this line please, I think  res.CoreConfigSchema().ImpliedType() comes from the current provider schema but instanceState.RawState may be on prev provider schema in the case of actual upgrade. We could try inferring the type from the value. I think it's about tradeoffs both approaches slightly problematic.

Reply:
We should be able to change that to 
```go
	json, err := ctyjson.Marshal(instanceState.RawState, res.RawState.Type())
```

I'll add a test where we do a more substantial upgrade to verify that this works as expected.

Reply:
It didn't work without the change. Good catch! I have added the fix and a test to go with it.